src/event/world/WorldSoundEvent.php

```diff
@@ -0,0 +1,81 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\event\world;
+
+use pocketmine\event\Cancellable;
+use pocketmine\event\CancellableTrait;
+use pocketmine\math\Vector3;
+use pocketmine\player\Player;
+use pocketmine\world\sound\Sound;
+use pocketmine\world\World;
+
+/**
+ * Called when a sound is played in a world using World->playSound()
+ * @see World::playSound()
+ */
+class WorldSoundEvent extends WorldEvent implements Cancellable{
+	use CancellableTrait;
+
+	/**
+	 * @param Player[]|null $players
+	 */
+	public function __construct(
+		World $world,
+		private Vector3 $position,
+		private Sound $sound,
+		private ?array $players = null
```

Comment:
why does this accept null?

Reply:
https://github.com/pmmp/PocketMine-MP/blob/stable/src/world/World.php#L661-L667

If players == null the world just sends it to the viewers. Developers may only want to change the sound and not the target, so why make them have to figure out who to send that sound to.

Reply:
I don't like that this can be null, it can be understood as that no player will hear the sound.

Reply:
What do we think about a field in the event such as ``toViewersOnly`` which essentially acts as if $players was null?

Reply:
The event would already provide the viewers, so your point doesn't make any sense.

Reply:
Currently, if no players are specified (``$players = null``) the server adds them to a packet buffer, and does not actually get the viewers of that chunk at that exact moment.

Reply:
Hmm, I see. That means that collecting the viewers at that point would change the internal buffering behaviour of sending sounds. Not sure what the best solution for that is.

Reply:
The solution that I used was if null was passed get the players near the sound before sending the event. The buffering behavior gets the players who can see that chunk either way.